packages/graph/hash_graph/lib/graph/src/store/postgres/query/compile.rs

```diff
@@ -0,0 +1,307 @@
+use std::{
+    borrow::Cow,
+    fmt::{Display, Formatter},
+    iter::once,
+    marker::PhantomData,
+};
+
+use postgres_types::ToSql;
+
+use crate::store::{
+    postgres::query::{
+        Column, ColumnAccess, Condition, EqualityOperator, Expression, Field, Function,
+        JoinExpression, Path, PostgresQueryRecord, SelectExpression, SelectStatement, Table,
+        TableAlias, TableName, Transpile, WhereExpression, WindowStatement, WithExpression,
+    },
+    query::{Filter, FilterExpression, Parameter},
+};
+
+pub struct CompilerArtifacts<'f> {
+    parameters: Vec<&'f (dyn ToSql + Sync)>,
+    current_alias: TableAlias,
+}
+
+pub struct SelectCompiler<'f, 'q, T> {
+    statement: SelectStatement<'q>,
+    artifacts: CompilerArtifacts<'f>,
+    _marker: PhantomData<fn(&'f T)>,
```

Comment:
Generics on `PhantomData` should match the main use case. of the generic. E.g. `PhantomData<T>` gives a hint to the user (and [dropck](https://doc.rust-lang.org/nomicon/dropck.html) but since [RFC 1238](https://rust-lang.github.io/rfcs/1238-nonparametric-dropck.html) this is not needed anymore). Strictly, this would be `T::Path`, not `T` but I would also be fine removing `fn`.

Reply:
Thanks for explaining.
So this means we would be using `SelectCompiler` in a situation where the T is used as a immutable borrow in some sort of method/function call?

Reply:
Yes 🙂 

As mentioned, it's not strictly a borrow of `T`, but I prefer `fn` here as we don't hold a reference to `T` (`T` is e.g. `DataType` or `ProeprtyType`)